internal/jimmhttp/rebac_admin/identities.go

```diff
@@ -93,19 +93,105 @@ func (s *identitiesService) UpdateIdentity(ctx context.Context, identity *resour
+	tuples, cNextToken, err := s.jimm.ListRelationshipTuples(ctx, user, apiparams.RelationshipTuple{
+		Object:       objUser.ResourceTag().String(),
+		Relation:     ofganames.AssigneeRelation.String(),
+		TargetObject: openfga.RoleType.String(),
+	}, int32(filter.Limit()), filter.Token()) // #nosec G115 accept integer conversion
+	if err != nil {
+		return nil, err
+	}
+
+	roles := make([]resources.Role, 0, len(tuples))
+	for _, t := range tuples {
+		dbRole, err := s.jimm.GetRoleManager().GetRoleByUUID(ctx, user, t.Target.ID)
+		if err != nil {
+			// Handle the case where the role was removed from the DB but a lingering OpenFGA tuple still exists.
+			// Don't return an error as that would prevent a user from viewing their groups, instead drop the role from the result.
+			if errors.ErrorCode(err) == errors.CodeNotFound {
+				continue
+			}
+			return nil, err
+		}
+		roles = append(roles, resources.Role{
+			Id:   &t.Target.ID,
+			Name: dbRole.Name,
+		})
+	}
+
+	originalToken := filter.Token()
+	res := resources.PaginatedResponse[resources.Role]{
+		Data: roles,
+		Meta: resources.ResponseMeta{
+			Size:      len(roles),
+			PageToken: &originalToken,
+		},
+	}
+	if cNextToken != "" {
+		res.Next.PageToken = &cNextToken
+	}
+	return &res, nil
 }
 
-// // PatchIdentityRoles performs addition or removal of a Role to/from an Identity.
+// PatchRoleIdentities performs addition or removal of identities to/from a Role identified by `roleId`.
 func (s *identitiesService) PatchIdentityRoles(ctx context.Context, identityId string, rolePatches []resources.IdentityRolesPatchItem) (bool, error) {
-	return false, v1.NewNotImplementedError("get identity roles not implemented")
+	user, err := utils.GetUserFromContext(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	objUser, err := s.jimm.FetchIdentity(ctx, identityId)
+	if err != nil {
+		return false, v1.NewNotFoundError(fmt.Sprintf("User with id %s not found", identityId))
+	}
+	additions := make([]apiparams.RelationshipTuple, 0)
+	deletions := make([]apiparams.RelationshipTuple, 0)
+	for _, p := range rolePatches {
+		if !jimmnames.IsValidRoleId(p.Role) {
+			return false, v1.NewValidationError(fmt.Sprintf("ID %s is not a valid role ID", p.Role))
+		}
+		t := apiparams.RelationshipTuple{
+			Object:       objUser.ResourceTag().String(),
+			Relation:     ofganames.AssigneeRelation.String(),
+			TargetObject: jimmnames.NewRoleTag(p.Role).String(),
+		}
+		if p.Op == resources.IdentityRolesPatchItemOpAdd {
+			additions = append(additions, t)
+		} else if p.Op == "remove" {
+			deletions = append(deletions, t)
+		}
+	}
+	if len(additions) > 0 {
+		err = s.jimm.AddRelation(ctx, user, additions)
+		if err != nil {
+			zapctx.Error(context.Background(), "cannot add relations", zap.Error(err))
+			return false, v1.NewUnknownError(err.Error())
+		}
+	}
+	if len(deletions) > 0 {
+		err = s.jimm.RemoveRelation(ctx, user, deletions)
```

Comment:
ah, ouch, this one could be tricky. what if user is member of a group that is an asignee of a role.. here we attempt to remove only direct relations, but `GetIdentityRoles` returnes indirect relations, too. 

in this case, should we verify if access was actually removed - identity no longer has asignee relation to the role - and report an error/suggestion that this might be a case of indirect relations?

Reply:
I think direct access is the right "view" for this handler. Indirect relation checks should be handled separately.

Reply:
There are separate handlers for assigning/removing groups from roles that I will make in a follow-up PR.